Gems/PhysX/Code/Source/PhysXCharacters/Components/CharacterGameplayComponent.cpp

```diff
@@ -93,25 +108,84 @@ namespace PhysX
 
     // CharacterGameplayRequestBus
     bool CharacterGameplayComponent::IsOnGround() const
+    {
+        if (m_cachedGroundState == CharacterGroundState::NotYetDetermined)
+        {
+            DetermineCachedGroundState();
+        }
+
+        return m_cachedGroundState == CharacterGroundState::Touching;
+    }
+
+    void CharacterGameplayComponent::DetermineCachedGroundState() const
     {
         Physics::Character* character = nullptr;
         Physics::CharacterRequestBus::EventResult(character, GetEntityId(), &Physics::CharacterRequests::GetCharacter);
         if (!character)
         {
-            return true;
+            m_cachedGroundState = CharacterGroundState::Touching;
+            return;
         }
 
         auto pxController = static_cast<physx::PxController*>(character->GetNativePointer());
         if (!pxController)
         {
-            return true;
+            m_cachedGroundState = CharacterGroundState::Touching;
+            return;
         }
 
+        // first check if we can use the character controller state, which should be cheaper than doing a scene query
+
+        // if the controller is slightly above an object or has not been asked to move downwards, the PxController may
+        // not report a touched actor or downward collision, so this can give false negatives, but should not give
+        // false positives, so it's useful as an early out
         physx::PxControllerState state;
         pxController->getState(state);
-        return
-            state.touchedActor != nullptr ||
-            (state.collisionFlags & physx::PxControllerCollisionFlag::eCOLLISION_DOWN) != 0;
+
+        if (state.touchedActor != nullptr || (state.collisionFlags & physx::PxControllerCollisionFlag::eCOLLISION_DOWN) != 0)
+        {
+            m_cachedGroundState = CharacterGroundState::Touching;
+            return;
+        }
+
+        // if we get to this point it's still unclear whether the character is touching the ground so
+        // use an overlap query to see if there's any geometry immediately below the character's foot position
+        if (auto* scene = character->GetScene())
+        {
+            // use a box shape for the overlap, even if the character geometry is a capsule, to avoid difficulties with
+            // the curved base of the capsule
+            AZ::Vector3 footBoxDimensions(0.0f, 0.0f, m_groundDetectionBoxHeight);
+            if (pxController->getType() == physx::PxControllerShapeType::eCAPSULE)
+            {
+                const float radius = static_cast<physx::PxCapsuleController*>(pxController)->getRadius();
+                footBoxDimensions.SetX(2.0f * radius);
+                footBoxDimensions.SetY(2.0f * radius);
```

Comment:
suggestion: Could we combine the Set in one call (not sure what the type is here by can we just do `Set(radius * 2.0f, radius * 2.0f);`

Reply:
It's a Vector3. Guess I could do ... = Vector3(2.0f * radius, 2.0f * radius, m_groundDetectionBoxHeight)

Reply:
I think that's a bit nice but totally up to you 👍